src/flag_gems/ops/randperm.py

```diff
@@ -0,0 +1,362 @@
+        key_block_rank = tl.where(key_digit_mask, key_block_rank - 1, 0)
+        bin_of_bucket = tl.sum(key_elem_mask)
+        partial_counter = bin_of_bucket | LOOKBACK_PARTIAL_MASK
+        tl.store(
+            d_lookback + p * grid1 * bins + pid1 * bins + bin_id,
+            partial_counter,
+            cache_modifier=".wt",
+        )
+        bin_offset = p * (bins + 1) + bin_id
+        prefix_offsets = tl.load(digit_hist + bin_offset)
+        bk = pid1 - 1
+        inc_sum = bin_of_bucket
+        while bk >= 0:
+            rd_lbk_offset = p * grid1 * bins + bk * bins + bin_id
+            partial_prefix = tl.load(d_lookback + rd_lbk_offset, volatile=True)
+            while partial_prefix == 0:
+                partial_prefix = tl.load(d_lookback + rd_lbk_offset, volatile=True)
+            inc_sum += (partial_prefix & LOOKBACK_VALUE_MASK).to(tl.int32)
+            if partial_prefix & LOOKBACK_GLOBAL_MASK:
+                # break
+                bk = -1
+            else:
+                bk -= 1
+        global_counter = inc_sum | LOOKBACK_GLOBAL_MASK
+        tl.store(
+            d_lookback + p * grid1 * bins + pid1 * bins + bin_id,
+            global_counter,
+            cache_modifier=".wt",
+        )
+        global_offsets = (prefix_offsets + inc_sum - bin_of_bucket) + key_block_rank
+        tl.store(key_out + global_offsets, key_data, mask=key_digit_mask)
+        tl.store(value_out + global_offsets, value_data, mask=key_digit_mask)
+
+
+# for parallelization, randomly shuffle the entire block rather than adjacent equal elements as pytorch cuda backend
+@libentry()
+@triton.jit(do_not_specialize=["philox_seed", "philox_offset"])
+def duplicate_keys_shuffle_kernel(
+    value_in, n_elements, philox_seed, philox_offset, BLOCK_SIZE: tl.constexpr
+):
+    pid0 = tl.program_id(0)
+    offset_range = tl.arange(0, BLOCK_SIZE)
+    value_offset = pid0 * BLOCK_SIZE + offset_range
+    value_mask = value_offset < n_elements
+    value_data = tl.load(value_in + value_offset, mask=value_mask)
+
+    philox_seed = philox_seed.to(tl.int64)
+    philox_offset = philox_offset.to(tl.int64)
+    c0 = (philox_offset & 0xFFFFFFFF).to(tl.uint32)
+    c1 = ((philox_offset >> 32) & 0xFFFFFFFF).to(tl.uint32)
+    i4 = tl.program_id(0) * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    c0 += i4
+    _O = c0 * 0
+    r0, _, _, _ = tl.philox(philox_seed, c0, c1, _O, _O)
+
+    _block_size = BLOCK_SIZE
+    r1 = r0 % _block_size.to(tl.uint32)
+    mask_val = _get_iinfo_val(tl.uint32, True)
+    r1 = tl.where(value_offset < n_elements, r1, mask_val)
+    _, sorted_chunk_index = argsort(r1, offset_range, 0, descending=False)
+    store_offset = pid0 * BLOCK_SIZE + sorted_chunk_index
+    tl.store(value_in + store_offset, value_data, mask=store_offset < n_elements)
+
+
+def sort_by_key(key, value):
+    n_elements = key.numel()
+    if n_elements > 2 * 1024:
+        # radix method
+        BLOCK_SIZE = 1024
+        bits_per_pass = 4
```

Comment:
How are these parameters determined? Do they need to be tuned?

Reply:
These parameters are set according to the perf of some test cases
Actually, **BLOCK_SIZE** and **bits_per_pass** is related to the shape of digit_hist&d_lookback which are intermediate tensors, so it looks can't find a way to extract the tuned value to create a tensor.